src/treelearner/kernels/histogram_16_64_256.hu

```diff
@@ -0,0 +1,188 @@
+/*!
+ * ibmGBT: IBM CUDA Accelerated LightGBM
+ *
+ * IBM Confidential
+ * Copyright (c) 2019 IBM Corporation. All rights reserved.
+ *
+ * The source code for this program is not published or otherwise
+ * divested of its trade secrets, irrespective of what has been
+ * deposited with the U.S. Copyright Office.
+ *
+ * US Government Users Restricted Rights - Use, duplication or
+ * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
+ */
+
+#ifndef _HISTOGRAM_16_64_256_KERNEL_
+#define _HISTOGRAM_16_64_256_KERNEL_
+
+//#pragma once
+
+// use double precision or not
+#ifndef USE_DP_FLOAT
+#define USE_DP_FLOAT 1
+#endif
+
+// ignore hessian, and use the local memory for hessian as an additional bank for gradient
+#ifndef CONST_HESSIAN
+#define CONST_HESSIAN 0
+#endif
+
+typedef unsigned char uchar;
+
+template<typename T>
+__device__ double as_double(const T t) {
+  static_assert(sizeof(T) == sizeof(double), "size mismatch");
+  double d; 
+  memcpy(&d, &t, sizeof(T)); 
+  return d;
+}
+template<typename T>
+__device__ ulong as_ulong(const T t) {
+  static_assert(sizeof(T) == sizeof(ulong), "size mismatch");
+  ulong u; 
+  memcpy(&u, &t, sizeof(T)); 
+  return u;
+}
+template<typename T>
+__device__ float as_float(const T t) {
+  static_assert(sizeof(T) == sizeof(float), "size mismatch");
+  float f; 
+  memcpy(&f, &t, sizeof(T)); 
+  return f;
+}
+template<typename T>
+__device__ uint as_uint(const T t) {
+  static_assert(sizeof(T) == sizeof(uint), "size_mismatch");
+  uint u; 
+  memcpy(&u, &t, sizeof(T)); 
+  return u;
+}
+template<typename T>
+__device__ uchar4 as_uchar4(const T t) {
+  static_assert(sizeof(T) == sizeof(uchar4), "size mismatch");
+  uchar4 u; 
+  memcpy(&u, &t, sizeof(T)); 
+  return u;
+}
+
+#if USE_DP_FLOAT == 1
+typedef double acc_type;
+typedef ulong acc_int_type;
+#define as_acc_type as_double
+#define as_acc_int_type as_ulong
+#else
+typedef float acc_type;
+typedef uint acc_int_type;
+#define as_acc_type as_float
+#define as_acc_int_type as_uint
+#endif
+
+// unroll the atomic operation for a few times. Takes more code space, 
+// but compiler can generate better code for faster atomics.
+#define UNROLL_ATOMIC 1
```

Comment:
I just looked in the .cu code and the .hu code, and yes, we can remove NVIDIA and UNROLL_ATOMIC and AMD_USE_DS_ADD_F32.  But I could not find any others that could be removed.  Since you said "many," I figured I should fess up and admit I couldn't find others...

Reply:
It looks good now, thanks!